beacon-chain/core/validators/validator.go

```diff
@@ -76,8 +77,11 @@ func InitiateValidatorExit(state *pb.BeaconState, idx uint64) (*pb.BeaconState,
 	if uint64(exitQueueChurn) >= churn {
 		exitQueueEpoch++
 	}
-	state.Validators[idx].ExitEpoch = exitQueueEpoch
-	state.Validators[idx].WithdrawableEpoch = exitQueueEpoch + params.BeaconConfig().MinValidatorWithdrawabilityDelay
+	validator.ExitEpoch = exitQueueEpoch
+	validator.WithdrawableEpoch = exitQueueEpoch + params.BeaconConfig().MinValidatorWithdrawabilityDelay
+	if err := state.UpdateValidatorAtIndex(idx, validator); err != nil {
+		return nil, err
+	}
 	return state, nil
```

Comment:
Return the updated validators list here instead of returning state. We will update all validators at the end of the state transition.